crates/accelerate/src/unitary_synthesis.rs

```diff
@@ -0,0 +1,1093 @@
+                        Some(false)
+                    } else {
+                        None
+                    }
+                }
+            }
+        }
+    };
+
+    if natural_direction == Some(true) && preferred_direction.is_none() {
+        return Err(QiskitError::new_err(format!(
+            concat!(
+                "No preferred direction of gate on qubits {:?} ",
+                "could be determined from coupling map or gate lengths / gate errors."
+            ),
+            qubits
+        )));
+    }
+
+    Ok(preferred_direction)
+}
+
+fn synth_su4_sequence(
+    su4_mat: &Array2<Complex64>,
+    decomposer_2q: &DecomposerElement,
+    preferred_direction: Option<bool>,
+    approximation_degree: Option<f64>,
+) -> PyResult<TwoQubitUnitarySequence> {
+    let is_approximate = approximation_degree.is_none() || approximation_degree.unwrap() != 1.0;
+    let synth = if let DecomposerType::TwoQubitBasisDecomposer(decomp) = &decomposer_2q.decomposer {
+        decomp.call_inner(su4_mat.view(), None, is_approximate, None)?
+    } else {
+        panic!("synth_su4_sequence should only be called for TwoQubitBasisDecomposer.")
+    };
+    let sequence = TwoQubitUnitarySequence {
+        gate_sequence: synth,
+        decomp_gate: decomposer_2q.gate.clone(),
+    };
+
+    match preferred_direction {
+        None => Ok(sequence),
+        Some(preferred_dir) => {
+            let mut synth_direction: Option<SmallVec<[u8; 2]>> = None;
+            // if the gates in synthesis are in the opposite direction of the preferred direction
+            // resynthesize a new operator which is the original conjugated by swaps.
+            // this new operator is doubly mirrored from the original and is locally equivalent.
+            for (gate, _, qubits) in &sequence.gate_sequence.gates {
+                if gate.is_none() || gate.unwrap().name() == "cx" {
+                    synth_direction = Some(qubits.clone());
+                }
+            }
+
+            match synth_direction {
+                None => Ok(sequence),
+                Some(synth_direction) => {
+                    let synth_dir = match synth_direction.as_slice() {
+                        [0, 1] => true,
+                        [1, 0] => false,
+                        _ => panic!(),
+                    };
+                    if synth_dir != preferred_dir {
+                        reversed_synth_su4_sequence(su4_mat, decomposer_2q, approximation_degree)
+                    } else {
+                        Ok(sequence)
+                    }
+                }
+            }
+        }
+    }
+}
+
+fn reversed_synth_su4_sequence(
+    su4_mat: &Array2<Complex64>,
+    decomposer_2q: &DecomposerElement,
+    approximation_degree: Option<f64>,
+) -> PyResult<TwoQubitUnitarySequence> {
+    let is_approximate = approximation_degree.is_none() || approximation_degree.unwrap() != 1.0;
+    let mut su4_mat_mm = su4_mat.clone();
+
```

Comment:
This is just a nitpick but maybe consider accepting an owned `Array2` instance in this case? Just do the cloning beforehand so it doesnt seem like you don't take ownership of this `Array2`.
```suggestion
fn reversed_synth_su4_sequence(
    su4_mat: Array2<Complex64>,
    decomposer_2q: &DecomposerElement,
    approximation_degree: Option<f64>,
) -> PyResult<TwoQubitUnitarySequence> {
    let is_approximate = approximation_degree.is_none() || approximation_degree.unwrap() != 1.0;
```

Reply:
I don't have a strong opinion, why not. Done in [6abcf43](https://github.com/Qiskit/qiskit/pull/13141/commits/6abcf43b32855d7eca73b19538c6dbfc11107b13) and  [27c3a0a](https://github.com/Qiskit/qiskit/pull/13141/commits/27c3a0a73787b0ee46df9d5c22cfafae2715549d)